Add a post about `--allow-undefined` and wasm targets (#1815)
rust-lang/blog.rust-lang.org: content/changes-to-webassembly-targets-and-handling-undefined-symbols.md

Reviewer: As a user of WebAssembly on the web, I’ve often had to deal with errors about a missing `env`, in situations where I accidentally introduce a dependency that tries to link to nonexistent system library symbols. These errors are hard to debug because they do not tell me which library is at fault or even (by default, with wasm-bindgen + wasm-pack glue code) which symbol.

> `Uncaught TypeError: Failed to resolve module specifier "env". Relative references must start with either "/", "./", or "../".`

I think it would be nice if this blog post mentioned that, with an example of such an error. That way, developers learn of how this change will help them, not just how it makes things more conventional.

+++
path = "2026/04/02/changes-to-webassembly-targets-and-handling-undefined-symbols"
title = "Changes to WebAssembly targets and handling undefined symbols"
authors = ["Alex Crichton"]
+++

Rust's WebAssembly targets are soon going to experience a change which has a
risk of breaking existing projects, and this post is intended to notify users of
this upcoming change, explain what it is, and how to handle it. Specifically all
WebAssembly targets in Rust have been linked using the `--allow-undefined` flag

> to `wasm-ld` historically, and this is being removed.

Reviewer: I think this reads better:

```suggestion
this upcoming change, explain what it is, and how to handle it. Specifically, all
WebAssembly targets in Rust have been linked using the `--allow-undefined` flag
to `wasm-ld`, and this is being removed.
```

}

fn init() {
    unsafe {
        mylibrary_init();
    }
}
```

The symbol `mylibrary_init` is an undefined symbol. This is typically defined by
a separate component of a program, such as an externally compiled C library,
which will provide a definition for this symbol. By passing `--allow-undefined`
to `wasm-ld`, however, it means that the above would generate a WebAssembly
module like so:

```wasm
(module
    (import "env" "mylibrary_init" (func $mylibrary_init))

    ;; ...
)
```

This means that the undefined symbol was ignored and ended up as an imported
symbol in the final WebAssembly module that is produced.

The precise history here is somewhat lost to time, but the current understanding
is that `--allow-undefined` was effectively required in the very early days of
introducing `wasm-ld` to the Rust toolchain. This historical workaround stuck
around til today and hasn't changed.

## What's wrong with `--allow-undefined`?

By passing `--allow-undefined` on all WebAssembly targets rustc is introducing
diverging behavior between other platforms and WebAssembly. The main risk of
`--allow-undefined` means that misconfiguration or mistakes in building can

> result in broken WebAssembly modules being produced as opposed to compilation

Reviewer: ```suggestion
By passing `--allow-undefined` on all WebAssembly targets, rustc is introducing
diverging behavior between other platforms and WebAssembly. The main risk of
`--allow-undefined` is that misconfiguration or mistakes in building can
result in broken WebAssembly modules being produced, as opposed to compilation
```

intentionally relying on this behavior. For example your application might have:

```rust
unsafe extern "C" {
    fn js_log(n: u32);
}

// ...
```

And then perhaps some JS code that looks like:

```js
let instance = await WebAssembly.instantiate(module, {
    env: {
        js_log: n => console.log(n),
    }
});
```

Effectively it's possible for users to explicitly rely on the behavior of
`--allow-undefined` generating an import in the final WebAssembly binary.

If users encounter this then the code can be fixed through a `#[link]` attribute
which explicitly specifies the `wasm_import_module` name:

```rust
#[link(wasm_import_module = "env")]
unsafe extern "C" {
    fn js_log(n: u32);
}

// ...
```

This will have the same behavior as before and will no longer be considered an
undefined symbol to `wasm-ld`, and it'll work both before and after this change.

Affected users can also compile with `-Clink-arg=--allow-undefined` as well to
quickly restore the old behavior.

## When is this change being made?

Removing `--allow-undefined` on wasm targets is being done in
[rust-lang/rust#149868]. That change is slated to land in nightly soon which

> will get released with Rust 1.96 on 2026-05-28. If you see any issues as a

Reviewer: ```suggestion
[rust-lang/rust#149868]. That change is slated to land in nightly soon, and will then get released with Rust 1.96 on 2026-05-28. If you see any issues as a
```

* If `mylibrary` was mistakenly not compiled and linked into a final
  application then the `mylibrary_init` symbol would end up imported rather than
  producing a linker error saying it's undefined.

* If external tooling is used to process a WebAssembly module, such as `wasm-bindgen` or `wasm-tools component new`, these tools don't know what to do with `"env"` imports by default and they are likely to provide an error message of some form that isn't clearly connected back to the original source code and where the symbols was imported from.

* For web users if you've ever seen an error along the lines of `Uncaught TypeError: Failed to resolve module specifier "env". Relative references must start with either "/", "./", or "../".` this can mean that `"env"` leaked into the final module unexpectedly and the true error is the undefined symbol error, not the lack of `"env"` items provided.

All native platforms consider undefined symbols to be an error by default, and
thus by passing `--allow-undefined` rustc is introducing surprising behavior on
WebAssembly targets. The goal of the change is to remove this surprise and
behave more like native platforms.

> ## What is going to break, and how to fix?

Author: I'm a bit confused, isn't that exactly what this section is talking about? This section was intended to address exactly what you're doing, implicitly generating imports from the `env` module, and how to retain that behavior.

Reviewer 2: Yeah, sorry, that was my bad. I misread. The first paragraph threw me off.
